### PR TITLE
Fix preview rendering for sharees

### LINKED ISSUE #826: [ERROR] Gallery does not display the right image but some old version instead
### Steps to reproduce

1. User1 uploads an image to OC into a shared folder
2. User2 changes the image
3. User1 wants to view the changed image > System shows original image uploaded by User 1

Downloading the file: Always shows the correct file. The update on the OC works properly. Also local sync works properly. It's just the gallery preview that is not showing the proper file version.

### Expected behaviour

Gallery must always show the real file, not some cached version of it.

### Actual behaviour

Gallery does not show the real file. It seems to display some cached version of it. We tried to delete Browser caches and we even tried with different browsers: The system shows the changed image when User2 is logged in and the non-changed image when User1 is logged in. Also restarting the Apache Webserver doesn't help. We even disabled the gallery and re-enabled it. Nothing. It always displays the originally uploaded file instead of the modified one.

### Server configuration

**Operating system**: Ubuntu 18.10

**Web server:** Apache/2.4.34 (Ubuntu)

**Database:** mysql  Ver 15.1 Distrib 10.1.38-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2

**PHP version:** PHP Version 7.2.15-0ubuntu0.18.10.1

**cloud server:** ownCloud

**cloud version:** 10.1.0 (stable)

**Updated from an older installation or fresh install:** 10.1.0 was a fresh install

**List of activated apps:**

<details>
<summary>App list</summary>

Enabled:
  - comments: 0.3.0
  - configreport: 0.1.2
  - dav: 0.4.0
  - federatedfilesharing: 0.3.1
  - federation: 0.1.0
  - files: 1.5.2
  - files_external: 0.7.1
  - files_pdfviewer: 0.10.0
  - files_sharing: 0.11.0
  - files_textviewer: 1.0.1
  - files_trashbin: 0.9.1
  - files_versions: 1.3.0
  - files_videoplayer: 0.9.9
  - firstrunwizard: 1.1.1
  - gallery: 16.1.1
  - market: 0.3.0
  - notifications: 0.4.1
  - provisioning_api: 0.5.0
  - systemtags: 0.3.0
  - updatenotification: 0.2.1
Disabled:
  - encryption
  - external
  - user_external
</details>
 
**The content of config/config.php:**

<details>
<summary>Config report</summary>

$CONFIG = array (
  'updatechecker' => false,
  'instanceid' => '1234',
  'trusted_domains' =>
  array (
    0 => 'mycloud.mydomain.tld',
  ),
  'datadirectory' => '/home/mycloud.mydomain.tld',
  'overwrite.cli.url' => 'https://mycloud.mydomain.tld',
  'dbtype' => 'mysql',
  'version' => '10.1.0.4',
  'dbname' => 'mycloud.mydomain.tld',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'mycloud.mydomain.tld',
  'dbpassword' => '1234',
  'logtimezone' => 'Europe/Zurich',
  'log_type' => 'owncloud',
  'loglevel' => 1,
  'logfile' => '/var/log/owncloud.log',
  'logdateformat' => 'M d H:i:s',
  'installed' => true,
  'mail_domain' => 'mydomain.tld',
  'mail_from_address' => 'data',
  'mail_smtpmode' => 'smtp',
  'mail_smtpsecure' => 'ssl',
  'mail_smtpauthtype' => 'PLAIN',
  'mail_smtpauth' => 1,
  'mail_smtphost' => 'mail.mydomain.tld',
  'mail_smtpport' => '465',
  'mail_smtpname' => 'mycloud.mydomain.tld',
  'mail_smtppassword' => '1234',

/* Performance settings */
'asset-pipeline.enabled' => true,
'memcache.local' => '\\OC\\Memcache\\Redis',
'filelocking.enabled' => 'true',
'memcache.distributed' => '\\OC\\Memcache\\Redis',
'memcache.locking' => '\\OC\\Memcache\\Redis',
'redis' =>
        array (
                'host' => 'localhost',
                'port' => 6379,
                'timeout' => 0,
                'dbindex' => 0,
        ),

/*'preview_max_scale_factor' => 1,
  'enabledPreviewProviders' =>
  array (
    0 => 'OC\\Preview\\PNG',
    1 => 'OC\\Preview\\JPEG',
    2 => 'OC\\Preview\\GIF',
    11 => 'OC\\Preview\\Illustrator',
    12 => 'OC\\Preview\\Postscript',
    13 => 'OC\\Preview\\Photoshop',
    14 => 'OC\\Preview\\TIFF',
    5 => 'OC\\Preview\\SVG'
  ),
*/
);
</details>

**Are you using external storage, if yes which one:** No

**Are you using encryption:** No

**Are you using custom gallery.cnf config files:** No

#### Web server error log

<details>
<summary>Web server error log</summary>
No error in any log. We see proper GET logs in the Apache access log. All good.
</details>

#### Cloud log
<details>
<summary>Cloud log</summary>
No error. 
</details>

### Client configuration

**Browser:** All sorts of: Firefox, Chrome, Opera, even IE11 > Same issue.

**Operating system:**

#### Browser log
a) The javascript console log
Deprecation warning: tipsy is deprecated. Use tooltip instead.
https://mycloud.mydomain.tld/core/js/js.js?v=7d6cf38bb9fca1af1c755d8f95eff063

b) The network log 
System is fetching this file:
https://mycloud.mydomain.tld/index.php/apps/gallery/preview/78721?width=3900&height=3900&c=2534722bd193e4f89b34706cb4e7f4b8&requesttoken=LzR4XW5JBiJgQjcbTVEBQA4aIFAVIDwhBxQCCi01BCU%3D%3AfEOm8pMvV5ZSzhSoytI2BVhwogqDclgdrfJrGBvl95A%3D

It's GET status 200
cache-control is set to: no-cache, must-revalidate
 
It seems to be all good. Checking this I think it COULD be, that the gallery is picking up the original file that is maybe in the trash. Because the file was deleted by the other user and replaced with the same filename. Perhaps the system refers to a file pointer instead of to the actual file in this directory.

-->

<details>
<summary>Browser log</summary>

Nothing to see. Console is empty.

</details>

### LINKED ISSUE #31855: Previews-thumbnails get out-of-sync when edited by sharer and sharee
### Steps to reproduce
1. as "user1" create folder "test" and in it put  "testfile.txt" with content "owner" - thumbnail says "owner" - good
2. share folder "test" with "user2"
3. as "user2", edit "testfile.txt" and put content "user2 text"

- thumbnail displayed to "user2" is not updated, it still says "owner"

4. login as "user1" - the thumbnail displays "user2 text" - good
5. login as  "user2" again - look at the thumbnail of the file

- thumbnail displayed to "user2" is not updated, it still says "owner"

6. login as "user1" again, edit the file and put content  "user1 text" - thumbnail is updated - good
7. login as  "user2" again - look at the thumbnail of the file

- thumbnail displayed to "user2" is not updated, it still says "owner"

### Expected behaviour
The share receiver "user2" should always see a preview-thumbnail that is in sync with the file content.

### Actual behaviour
The share receiver "user2" gets "stuck" with the preview-thumbnail that was generated at the time they first received/looked at the file.

### Server configuration
10.0.9RC1 plus PR #31853 (which fixes missing hooks for thumbnail generation in 10.0.9RC1)

10.0.8 as symptoms similar to this, but might be a bit different.

### LINKED ISSUE #39202: Icon/Preview of image file not updated in non-owner WebUI shared folders after updating the file
### Steps to reproduce
1. Run ownCloud docker container as 'admin' user.
2. Create another user, e.g. 'Tester'.
3. Create and share folder with 'Tester'.
4. Place an image file (jpg or png) inside the shared folder.
5. Use WebUI to login as 'Tester' and observe an image file in the shared folder.
6. Logout as 'Tester' from WebUI.
7. Update the image file with some different content, but using the same file name (replace the file).
8. Use WebUI to login as 'Tester' and observe an image file in the shared folder.

### Expected behaviour
Expected to see the file icon and preview to reflect the updated content.

### Actual behaviour
The file icon and preview show the previous content, but size and upload are correct.

### Server configuration
**Operating system**: Ubuntu 20.04

**Web server:** Caddy as reverse proxy (also tested on Plesk with Nginx)

**Database:** ownCloud default

**PHP version:** 7.4.3

**ownCloud version:** 10.8.0 (stable)

**Updated from an older ownCloud or fresh install:** Fresh install

**Where did you install ownCloud from:** owncloud/server:latest from Docker Hub

**Signing status (ownCloud 9.0 and above):**

```
Login as admin user into your ownCloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results into https://gist.github.com/ and puth the link here.
```

**The content of config/config.php:**

```
Log in to the web-UI with an administrator account and click on
'admin' -> 'Generate Config Report' -> 'Download ownCloud config report'
This report includes the config.php settings, the list of activated apps
and other details in a well sanitized form.

or 

If you have access to your command line run e.g.:
sudo -u www-data php occ config:list system
from within your ownCloud installation folder

*ATTENTION:* Do not post your config.php file in public as is. Please use one of the above
methods whenever possible. Both, the generated reports from the web-ui and from occ config:list
consistently remove sensitive data. You still may want to review the report before sending.
If done manually then it is critical for your own privacy to dilligently
remove *all* host names, passwords, usernames, salts and other credentials before posting.
You should assume that attackers find such information and will use them against your systems.
```

**List of activated apps:** ownCloud default plus Wallpaper

```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your ownCloud installation folder.
```

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Firefox 92.0 and/or Google Chrome 93.0.4577.63

**Operating system:** Windows 10

#### Web server error log
```
Insert your webserver log here
```

#### ownCloud log (data/owncloud.log)
```
Insert your ownCloud log here
```

#### Browser log
```
Insert your browser log here, this could for example include:

a) The javascript console log
b) The network log 
c) ...
```

### PR DESCRIPTION
**Pull request by JammingBen**
## Description
Previous to this fix, the previews of updated shared files did not change for sharees. Those previews now get re-generated if the content of the files changed.

## Related Issue
- Fixes https://github.com/owncloud/core/issues/39202
- Fixes https://github.com/owncloud/core/issues/31855
- Fixes https://github.com/owncloud/gallery/issues/826

## How Has This Been Tested?
See steps in https://github.com/owncloud/core/issues/39202

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by phil-davis** [comment]
- admin shares `text.txt` with Alice
- Alice edits `text.txt` (for example, online with files_texteditor)
- the preview for Alice is not updated
- Alice does ctrl-F5 to refresh, or logs out and back in again - still the preview is not updated
- admin logs in - the updated preview is shown
- Alice logs in - now the updated preview is shown

In the combinations that I tried, it seems that the sharer needs to login and have the file displayed, and then the preview gets updated for everyone.

**Comment by JanAckermann** [comment]
@phil-davis, we was always logged with the sharer, so yes could be an issue :/

**Comment by JammingBen** [comment]
This happens because the preview of the file owner gets deleted when a sharee edits a file. Therefore we don't have a checksum to compare. @phil-davis Please have a look again, should work now.

**Comment by phil-davis** [comment]
> This happens because the preview of the file owner gets deleted when a sharee edits a file. Therefore we don't have a checksum to compare. @phil-davis Please have a look again, should work now.

Works now. I tried combinations of:

- share a single file with a group that has Alice and Brian in it. Alice edits the file, both Brian and admin see the updated preview.
- share a single file with each of Alice and Brian individually. Alice edits the file, both Brian and admin see the updated preview.
- share a folder with a group that has Alice and Brian in it. Alice edits a file in the folder, both Brian and admin see the updated preview.
- share a folder with each of Alice and Brian individually. Alice edits a file in the folder, both Brian and admin see the updated preview.

LGTM

**Review by phil-davis — APPROVED** [review]
LGTM - it should be easy to add a few unit tests to cover the new block of code???

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUIComments-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32252/124/1

File: lib/private/Preview.php [hunk 1]
```diff
@@ -763,6 +764,33 @@ public function getPreview() {
...
 			$this->getCachedPreview($cached);
+
+			if ($this->preview) {
+				$mount = $fileInfo->getMountPoint();
+
+				// Updating a shared file will always re-generate the preview for the
+				// share owner, but not its sharees. Therefore we need to check if the
+				// current cached preview is different from the share owner's one.
+				if ($mount instanceof SharedMount) {
+					$owner = $fileInfo->getOwner();
```
Comment (jvillafanez) [inline lib/private/Preview.php:775]: does this work for external storages?

Comment (JammingBen) [reply lib/private/Preview.php:775]: I tried with sharing an external storage as a whole and by sharing a file inside an external storage, both worked. Is that what you meant?

**Review by jvillafanez — APPROVED** [review]

**Comment by phil-davis** [comment]
Note: I created issue #39211 to add some acceptance tests. It should not be too hard to keep a record of some simple previews in the acceptance test fixtures, and compare them to the actual generated previews as files are changed.

**Pull request merged**